common/flatpak-dir.c

```diff
@@ -4726,50 +4753,72 @@ flatpak_dir_pull (FlatpakDir                           *self,
 
           context = flatpak_main_context_new_default ();
 
-          ostree_repo_find_remotes_async (self->repo, (const OstreeCollectionRef * const *) collection_refs_to_fetch,
+          ostree_repo_find_remotes_async (repo, (const OstreeCollectionRef * const *) collection_refs_to_fetch,
                                           find_options,
                                           NULL /* default finders */, progress, cancellable,
                                           async_result_cb, &find_result);
 
           while (find_result == NULL)
             g_main_context_iteration (context, TRUE);
 
-          allocated_results = ostree_repo_find_remotes_finish (self->repo, find_result, error);
+          allocated_results = ostree_repo_find_remotes_finish (repo, find_result, error);
 
           results = (const OstreeRepoFinderResult * const *) allocated_results;
           if (results == NULL)
-            return FALSE;
+            goto out;
 
-          for (i = 0, rev = NULL; results[i] != NULL && rev == NULL; i++)
-            rev = g_strdup (g_hash_table_lookup (results[i]->ref_to_checksum, &collection_ref));
+          /* Go ahead and pull commit metadata now, so that if any of the
+           * signatures are bad we can fall back to another rev from another
+           * result. This also means the flatpak_dir_setup_extra_data() call
+           * below won't need to fetch the metadata. */
+          {
+            OstreeRepoPullFlags metadata_pull_flags = flags | OSTREE_REPO_PULL_FLAGS_COMMIT_ONLY;
+            /*TODO: Use OSTREE_REPO_PULL_FLAGS_MIRROR for all collection-ref pulls */
+#if OSTREE_CHECK_VERSION (2019, 2)
+            metadata_pull_flags |= OSTREE_REPO_PULL_FLAGS_MIRROR;
+#endif
+            if (!repo_pull (repo, state->remote_name,
+                            NULL, ref, NULL, results,
+                            flatpak_flags, metadata_pull_flags,
+                            progress, cancellable, error))
+              {
+                g_prefix_error (error, _("While pulling %s from remote %s: "), ref, state->remote_name);
+                goto out;
+              }
+          }
+
+          if (!flatpak_repo_resolve_rev (repo, collection_ref.collection_id,
+                                         state->remote_name, ref, TRUE, &rev,
+                                         cancellable, error))
+            goto out;
 
           if (rev == NULL)
-            return flatpak_fail_error (error, FLATPAK_ERROR_INVALID_DATA, _("No such ref (%s, %s) in remote %s or elsewhere"),
-                                       collection_ref.collection_id, collection_ref.ref_name, state->remote_name);
+            {
+              g_set_error (error, FLATPAK_ERROR, FLATPAK_ERROR_INVALID_DATA, _("No such ref (%s, %s) in remote %s or elsewhere"),
+                           collection_ref.collection_id, collection_ref.ref_name, state->remote_name);
+              goto out;
```

Comment:
This control path could fall through to the `g_assert(); goto out` below.

Reply:
That might be too unobvious to someone reading it in future though.

Reply:
Yeah, I pushed a fixup to get rid of the `if (rev == NULL)` block that's now redundant.